
# HG changeset patch
# User welpy-cw <h.w.forms@arcor.de>
# Date 1765293149 0
# Node ID 6d976af71db8a4c1491744ffbe30df416dd98cfc
# Parent  8f24276c73cd79b7de18466a67326a646844f279
Bug 1877646 - Exclude saved searches from being added to unified folders. r=darktrojan


When a smart folder was created, virtual subfolders were included even though
they could not be part of the underlying search session. These subfolders did
not appear in the smart folder's properties; however, they were included in the
displayed count of selected folders. Conversely, when a new saved search was
created as a subfolder of a smart folder, it only appeared as part of the
remaining folders. The same thing happened when a saved search was renamed.

To align these behaviors, I suggest not adding virtual folders to smart folders
and letting them be part of an account's remaining folders in unified folder
mode. This will also prevent the IMAP error alert described in this bug.

Differential Revision: https://phabricator.services.mozilla.com/D248709

diff --git a/mail/modules/SmartMailboxUtils.sys.mjs b/mail/modules/SmartMailboxUtils.sys.mjs
--- a/mail/modules/SmartMailboxUtils.sys.mjs
+++ b/mail/modules/SmartMailboxUtils.sys.mjs
@@ -17,17 +17,18 @@ const folderTypes = [
   { flag: Ci.nsMsgFolderFlags.Drafts, name: "Drafts", type: "drafts" },
   { flag: Ci.nsMsgFolderFlags.Templates, name: "Templates", type: "templates" },
   { flag: Ci.nsMsgFolderFlags.SentMail, name: "Sent", type: "sent" },
   { flag: Ci.nsMsgFolderFlags.Archive, name: "Archives", type: "archives" },
   { flag: Ci.nsMsgFolderFlags.Junk, name: "Junk", type: "junk" },
   { flag: Ci.nsMsgFolderFlags.Trash, name: "Trash", type: "trash" },
   // { flag: Ci.nsMsgFolderFlags.Queue, name: "Outbox", type: "outbox" },
 ];
-const allFlags = folderTypes.reduce((all, cur) => all | cur.flag, 0);
+const allSpecialFolderFlags =
+  Ci.nsMsgFolderFlags.SpecialUse | Ci.nsMsgFolderFlags.Virtual;
 
 class SmartMailbox {
   #tagsFolder = null;
   #rootFolder = null;
   #server = null;
   #account = null;
   #TagFolderURIs = new Map();
 
@@ -169,19 +170,19 @@ class SmartMailbox {
             })
           );
         }
         if (!subFolders?.length) {
           return;
         }
 
         for (const sf of subFolders) {
-          // Add all of the subfolders except the ones that belong to
+          // Add all real subfolders except the ones that belong to
           // a different folder type.
-          if (!(sf.flags & allFlags)) {
+          if (!(sf.flags & allSpecialFolderFlags)) {
             searchFolders.push(sf);
             recurse(sf);
           }
         }
       };
 
       for (const server of lazy.MailServices.accounts.allServers) {
         for (const f of server.rootFolder.getFoldersWithFlags(

